Add upgrader related logic to flatbuffer

qihqi opened the pull request:
Summary: title

Test Plan: unittest

Differential Revision: D33593056

tugsbayasgalan commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -146,6 +152,13 @@ std::unique_ptr<mobile::Function> FlatbufferLoader::parseFunction(
+  appendUpgraderFunctions(function.get());
+  // 2. Decides if upgrader is needed
+  const uint32_t operator_version = module_->version();
+  bool use_upgrader =
```
I believe this is not necessary because the model version cannot go beyond kProducedFileFormatVersion

qihqi replied:
I copied from here: https://github.com/pytorch/pytorch/blob/c53b3ed20fc84438b5354bc81d410b4d1a1c4e4f/torch/csrc/jit/mobile/import.cpp#L389

I guess it's possible because the binary writing the file and the binary reading the file are not the same version.

tugsbayasgalan commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -320,9 +321,15 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
+#if ENABLE_UPGRADERS
+  const uint32_t version = caffe2::serialize::kMinProducedFileFormatVersion;
+#else
+  const uint32_t version = caffe2::serialize::kProducedFileFormatVersion;
```
is the version always be either of these two no matter what operator the model has?

qihqi replied:
Q: What is the correct way to find operator version?

tugsbayasgalan commented on torch/csrc/jit/frontend/versioned_symbols.cpp:
```diff
@@ -106,5 +108,42 @@ uint64_t get_min_version_for_kind(const NodeKind& kind) {
+uint64_t get_min_operator_version(const Module& module) {
+#if ENABLE_UPGRADERS
+  uint64_t min_version = caffe2::serialize::kMinSupportedFileFormatVersion;
+#else
```
you can probably remove this macro now that its been a while we enabled this feature and got no issues so far.

qihqi replied:
done.

tugsbayasgalan commented on torch/csrc/jit/frontend/versioned_symbols.cpp:
```diff
@@ -106,5 +108,42 @@ uint64_t get_min_version_for_kind(const NodeKind& kind) {
+    // are not covered by the new upgraders
+    min_version =
+        std::max(min_version, get_min_version_for_kind(node->kind()));
+#else
```
same here

qihqi replied:
done

tugsbayasgalan commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -743,33 +743,9 @@ struct PythonPrintImpl {
     }
   }
 
-  void checkVersion(Node* node) {
```
Did you check how this function is called? I remember it does some complicated traverse of the graph to find each node. I could be wrong tho

tugsbayasgalan replied:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/python_print.cpp#L1255 like this

qihqi replied:
this bit is reverted

tugsbayasgalan commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -351,10 +352,15 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
   auto mod = CreateModule(
       fbb,
-      0, /* version */
+      version,
```
How is this used in the new code?

tugsbayasgalan commented on torch/csrc/jit/mobile/compatibility/model_compatibility.cpp:
```diff
@@ -411,5 +413,27 @@ ModelCompatCheckResult is_compatible(
+  uint64_t min_version = caffe2::serialize::kMinSupportedFileFormatVersion;
+  for (const auto& func : module.compilation_unit().methods()) {
+    for (const auto& op_name : func->get_code().op_names_) {
+      auto schema_name = op_name.name + "." + op_name.overload_name;
```
```suggestion

if (op_name.overload_name != "") {
     auto schema_name = op_name.name + "." + op_name.overload_name;
} else {
     auto schema_name = op_name.name
}

```

tugsbayasgalan replied:
```suggestion

if (op_name.overload_name != "") {
     auto schema_name = op_name.name + "." + op_name.overload_name;
} else {
     auto schema_name = op_name.name
}

```

tugsbayasgalan commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -222,6 +222,7 @@ class BytecodeDeserializer final {
   // operators from the given model. If it's less than the current runtime,
   // upgrader will be applied at loading stage.
   uint64_t operator_version_;
+  uint64_t model_version_ = 3; // start at 3
```
Can this be kMinProducedFileFormatVersion?

cccclai replied:
not related to this change, `bytecode_version` is a better name after we introduce `operator_version`...

tugsbayasgalan commented on torch/csrc/jit/mobile/compatibility/model_compatibility.cpp:
```diff
@@ -411,5 +413,27 @@ ModelCompatCheckResult is_compatible(
+  }
+}
+
+uint64_t get_min_operator_version_via_bytecode(const mobile::Module& module) {
```
Can we rename this to something else? This version is calculated based on operator version map, not specifically for bytecodes.

cccclai replied:
Yeah agree. It's not specifically for bytecodes.

tugsbayasgalan commented on torch/csrc/jit/mobile/compatibility/model_compatibility.cpp:
```diff
@@ -411,5 +413,27 @@ ModelCompatCheckResult is_compatible(
+  }
+}
+
+uint64_t get_min_operator_version_via_bytecode(const mobile::Module& module) {
```
Can we rename this to something else? This version is calculated based on operator version map, not specifically for bytecodes.

tugsbayasgalan commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -226,7 +232,14 @@ std::unique_ptr<mobile::Function> FlatbufferLoader::parseFunction(
 
   std::unordered_set<std::string> unsupported_op_names;
-  const int64_t model_version = 0x6L;
+  const int64_t model_version = module_->version();
```
so this is at 9 now?

tugsbayasgalan commented:
From server side it looks OK. @cccclai Could you review from mobile side?

cccclai commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -294,12 +295,6 @@ void BytecodeDeserializer::parseFunctionSchema(
   }
 }
 
-void BytecodeDeserializer::init_upgrader(mobile::Function* function) {
```
Ok looks like we are trying to re-use this code...is there a better place to put re-use code between `import.cpp` and `flatbuffer_load.cpp`?

cccclai commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -377,6 +379,9 @@ mobile::Module jitModuleToMobile(
   m.setDebugTable(MobileDebugTable(
       debug_handle_cs_ptr_map.begin(), debug_handle_cs_ptr_map.end()));
+
+  m.set_min_operator_version(get_min_operator_version_via_bytecode(m));
+  m.set_bytecode_version(options.model_version);
```
The operator version is set when exporting the model, and it's not only for bytecode model. We probably don't need to set it here? cc: @tugsbayasgalan

cccclai commented on torch/csrc/jit/serialization/mobile_bytecode.fbs:
```diff
@@ -186,7 +186,7 @@ table ExtraFile {
 
 table Module {
-  version:int;
+  version:int;  // overall file version, will start at 9
```
Looks like duplicate? what is this version?

cccclai replied:
If it start at 9, it probably just means `operator_version`..

cccclai commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -364,10 +390,18 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
   for (const auto& ival : jit_constants) {
     jit_constants_indexes.emplace_back(storeIValueAndGetIndex(fbb, ival));
   }
+  const uint32_t version =
+      kMinVersion > caffe2::serialize::kProducedBytecodeVersion
+      ? kMinVersion
+      : caffe2::serialize::kProducedBytecodeVersion;
```
Looks like it's not needed
```suggestion

```

cccclai commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -364,10 +390,18 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
   auto mod = CreateModule(
       fbb,
-      0, /* version */
+      version,
```
Seems not needed.
```suggestion
```

qihqi replied:
kProducedBytecodeVersion is 8 currently. Changing it to 9 now will also change the pickle files generated.

cccclai reviewed (commented):
Overall looks great! Thanks for the effort to re-use code and improve the code quality. A few major comments
- mobile runtime doesn't include `model_compatibility.cpp`. try to move the re-use code to somewhere else shared by both `import.cpp` and `flatbuffer_load.cpp`. Create one if not exists.
- `versions` field is not needed in schema. Currently there are two versions, `bytecode_version` and `operator_version` and they are independent to each other.
- Looks like this pr combines some changes not quite related to upgrader. Seems like they are mostly about jit source serialization/deserailization. Can we decouple the change?

qihqi reviewed (commented):
Hi @cccclai addressing the bullet points
* `model_compatiliby.h` changed to `import_export_functions.h`
* `version` is meant to give to backport and will start at 9. bytecode version is given to parseOperators; operator version is given to upgraders. 
* Spurious changes are from a different diff in a diff stack, which are landed now. They should disappear now.

cccclai reviewed (commented):
Thanks for moving them to `import_export_functions.cpp`. Two major comments:

- Can we re-use bytecode version?
- Depends on @tugsbayasgalan 's answer, can we re-use the calculation for max of min op version?

qihqi commented:
@cccclai Clarification on version: As of the moment after this PR lands. if you call `_save_for_mobile` you get a pickle file with bytecode_version=8, op_version=3 (or whatever it happen to compute). If you call `save_mobile_module` you get a flatbuffer file with version=9, bytecode_version=8, op_version=3.

Why can't we have just 2 versions as before? Suppose we do, then either we do:
* bytecode_version=9, op_version=3: that won't pass unit test because bytecode_version is given to parseOperators; note that the pickle and flatbuffer actually have the same bytecodes in memory; but one is handled by calling parseOperator(..., 8) one is handled by parseOperator(..., 9) so they yield different result.
* bytecode_version=9, op_version=3. Then backport system cannot distinguish between pkl and flatbuffer anymore.

This is an artifact of needing to have 2 fileformats in the same time. In the future we can reconcile.

tugsbayasgalan commented:
> @cccclai Clarification on version: As of the moment after this PR lands. if you call `_save_for_mobile` you get a pickle file with bytecode_version=8, op_version=3 (or whatever it happen to compute). If you call `save_mobile_module` you get a flatbuffer file with version=9, bytecode_version=8, op_version=3.
> 
> Why can't we have just 2 versions as before? Suppose we do, then either we do:
> 
> * bytecode_version=9, op_version=3: that won't pass unit test because bytecode_version is given to parseOperators; note that the pickle and flatbuffer actually have the same bytecodes in memory; but one is handled by calling parseOperator(..., 8) one is handled by parseOperator(..., 9) so they yield different result.
> * bytecode_version=8, op_version=3. Then backport system cannot distinguish between pkl and flatbuffer anymore.
> 
> This is an artifact of needing to have 2 fileformats in the same time. In the future we can reconcile.

Can we name the version to something like flatbuffer_version? I agree that just saying version is little confusing

cccclai commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -383,9 +388,12 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
+      static_cast<uint32_t>(module.min_operator_version());
+  uint32_t bytecode_version = static_cast<uint32_t>(module.bytecode_version());
+  if (bytecode_version < kMinVersion) {
+    bytecode_version = kMinVersion;
```
I'm a bit confused by this logic here. Any reason to set bytecode_version to 9?

cccclai reviewed (commented):
Thanks for identifying the unnecessary dependency of parsing operators on bytecode version and rebasing! Now It looks nice. Just one minor feedback

suo commented:
if you want the linter to skip these files, configure the appropriate exclude patterns in `.lintrunner.toml`.

qihqi committed:
Add upgrader related logic to flatbuffer (#71451)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71451

title

Test Plan: unittest

Reviewed By: tugsbayasgalan

Differential Revision: D33593056

fbshipit-source-id: 3f2afc7525a6f9c2cfc21b6006164383624c9d07

malfet commented:
@pytorchbot revert this got fix syncbranches (i.e. allow cherry-pick of https://github.com/pytorch/pytorch/commit/8e09e20c1dafcdbdb45c2d1574da68a32e54a3a5)
